cachi2/core/package_managers/rpm/main.py

```diff
@@ -31,6 +33,124 @@
+    def from_filepath(cls, rpm_filepath: Path, rpm_download_metadata: dict[str, Any]) -> "Package":
+        """Instantiate a package dataclass instance from a download RPM file path."""
+        kwargs: dict[str, Optional[str]] = {}
+        kwargs.update(cls._query_rpm_fields(rpm_filepath))
+
+        repoid = rpm_download_metadata.get("repoid")
+        is_srpm = rpm_filepath.name.endswith("src.rpm")
+
+        if is_srpm:
+            # Red Hat PURL RPM guideline suggests injecting 'src' into the arch qualifier for SRPMS
+            kwargs["arch"] = "src"
+
+        kwargs["repository_id"] = repoid if repoid and not repoid.startswith("cachi2") else None
+        kwargs["download_url"] = rpm_download_metadata["url"]
+        kwargs["checksum"] = rpm_download_metadata.get("checksum")
+
+        # Disable mypy here:
+        # - the required fields here correspond with mandatory RPM tags, so they won't be None
+        # - download_url isn't an RPM tag, but it's non-null value is guarded by a pydantic model
+        package = cls(**kwargs)  # type: ignore
+        log.debug("RPM package attributes for '%s': %s", rpm_filepath, package)
+        return package
+
+    @staticmethod
+    def _query_rpm_fields(file_path: Path) -> dict[str, str]:
+        """Query a set of RPM tags.
+
+        Tags which are optional and not set won't be returned in the resulting dict.
+        """
+        ret = {}
+        query_format = (
+            # all nvra macros should be present/mandatory in RPM
+            "name=%{NAME}\n"
+            "version=%{VERSION}\n"
+            "release=%{RELEASE}\n"
+            "arch=%{ARCH}\n"
+            # vendor and epoch are optional RPM tags; return "" if not set instead of "(None)"
+            "vendor=%|VENDOR?{%{VENDOR}}:{}|\n"
+            "epoch=%|EPOCH?{%{EPOCH}}:{}|\n"
+        )
+        rpm_args = [
+            "-q",
+            "--queryformat",
+            query_format.strip(),
+            str(file_path),
+        ]
+        rpm_fields = run_cmd(cmd=["rpm", *rpm_args], params={})
+        for entry in rpm_fields.split("\n"):
+            key, value = entry.partition("=")[::2]
+            if not value:
+                continue
+            ret[key] = value
+
+        return ret
+
+    @property
+    def purl(self) -> str:
+        """Get the purl for this package."""
+        # TODO: get rid of these mappings the moment the upstream PURL spec provides clear
+        # guidelines where does the namespace value come from, i.e. not the VENDOR RPM header tag
+        vendor_namespace_mapping = {
+            "Red Hat": "redhat",  # common Vendor string 'Red Hat Inc.'
+            "Fedora": "fedora",  # common Vendor string: Fedora Project, Fedora Copr - group @XYZ
+            "SUSE": "suse",  # common Vendor string: SUSE LLC <https://www.suse.com/>
+        }
+
+        qualifier_fields = {
+            "epoch": self.epoch,
+            "arch": self.arch,
+            "repository_id": self.repository_id,
```

Comment:
This is a very "cornerish" case, but since `repoid` is a open string field in the `rpms.lock.yaml` file, one could set it to something such as `myrepo&iamevil=true`, which will end up looking like another qualifier in the PURL. Do you think it's worth doing a basic santization of this field for the PURL generation purpose?

Reply:
No, because, while admittedly it might look nicer in the PURL, such a repo ID would be invalid. Not only that, the project would fail the DNF install part:
```
$ cat /etc/yum.repos.d/fedora.repo
[asdf&foo=bar]
name=Fedora $releasever - $basearch
#baseurl=http://download.example/pub/fedora/linux/releases/$releasever/Everything/$basearch/os/
metalink=https://mirrors.fedoraproject.org/metalink?repo=fedora-$releasever&arch=$basearch
...

$ dnf install zzuf
Bad id for repo: asdf&foo=bar, byte = & 4
Last metadata expiration check: 1:42:04 ago on Wed 21 Aug 2024 15:15:16 CEST.
No match for argument: zzuf
Error: Unable to find a match: zzuf
```
as you can see the repo got completely ignored. Relevant piece of the [docs](https://dnf.readthedocs.io/en/latest/conf_ref.html#description):

> The allowed characters of repo ID string are lower and upper case alphabetic letters, digits, -, _, . and :.

I'm playing with the idea to actually check this with a regex, but then again, it's clearly a configuration error that originates outside of cachi2 realm, so I'd say we don't do anything and we'll let it fail. As long as it fails, nobody even looks at PURLs :) .

Reply:
For what it's worth, `PackageUrl` does encode qualifiers in accordance with the purl spec. We should use it.